src/manager/checkpoint/manager/topdown.rs

```diff
@@ -1,19 +1,90 @@
-use std::fmt::{Display, Formatter};
 // Copyright 2022-2023 Protocol Labs
 // SPDX-License-Identifier: MIT
+use std::fmt::{Display, Formatter};
+
 use crate::manager::checkpoint::CheckpointManager;
 use fvm_shared::address::Address;
 use fvm_shared::clock::ChainEpoch;
 use ipc_sdk::subnet_id::SubnetID;
 
-pub struct TopDownCheckpointManager;
 use crate::config::Subnet;
 use crate::lotus::client::DefaultLotusJsonRPCClient;
+use crate::lotus::message::ipc::IPCReadGatewayStateResponse;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
 use async_trait::async_trait;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use fvm_shared::MethodNum;
+use ipc_gateway::TopDownCheckpoint;
+
+pub struct TopDownCheckpointManager {
+    parent: Subnet,
+    parent_client: DefaultLotusJsonRPCClient,
+    child_subnet: Subnet,
+    child_client: DefaultLotusJsonRPCClient,
+
+    checkpoint_period: ChainEpoch,
+}
+
+impl TopDownCheckpointManager {
+    pub async fn new(
+        parent_client: DefaultLotusJsonRPCClient,
+        parent: Subnet,
+        child_client: DefaultLotusJsonRPCClient,
+        child_subnet: Subnet,
+    ) -> anyhow::Result<Self> {
+        let checkpoint_period = obtain_checkpoint_period(&child_subnet.id, &child_client).await?;
+        Ok(Self {
+            parent,
+            parent_client,
+            child_subnet,
+            child_client,
+            checkpoint_period,
+        })
+    }
+}
 
 impl Display for TopDownCheckpointManager {
-    fn fmt(&self, _f: &mut Formatter<'_>) -> std::fmt::Result {
-        todo!()
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "top-down, parent: {:}, child: {:}",
+            self.parent.id, self.child_subnet.id
+        )
+    }
+}
+
+impl TopDownCheckpointManager {
+    async fn sync_subnet_gateway_state(&self) -> anyhow::Result<IPCReadGatewayStateResponse> {
+        let child_head = self.child_client.chain_head().await?;
+        let cid_map = child_head.cids.first().unwrap().clone();
+        let child_tip_set = Cid::try_from(cid_map)?;
+
+        self.child_client
+            .ipc_read_gateway_state(child_tip_set)
+            .await
+    }
+
+    async fn parent_tipset(&self) -> anyhow::Result<Cid> {
```

Comment:
This function doesn't return the tipset but its key, what about renaming it to:
```suggestion
    async fn parent_head(&self) -> anyhow::Result<Cid> {
```